Adding Sanitizer interface and implementation

longquanzheng opened the pull request:
This is replacement of https://github.com/uber-java/tally/pull/70

* Sanitizes metric names and tag keys and values.
* Does nothing by default.
* The ScopeBulider can be configured with the standard M3Sanitizer.
* Fixes #8

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber-java/tally?pullRequest=97) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/uber-java/tally?pullRequest=97) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: ravirajj<br/>:x: longquanzheng<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber-java/tally?pullRequest=97) it.</sub>

longquanzheng commented:
@alexeykudinkin Hi, can you take a look? I have addressed the comments from the previous PR. LMK if you have any others.

longquanzheng commented:
looks like unit tests are failing. Working on fixing

longquanzheng commented:
@alexeykudinkin Hi, can you take a look? I have fixed all the tests. Thanks in advance.

longquanzheng commented:
@alexeykudinkin @andrewmains12 Hello, sorry to bother you guys again. I would really appreciate if you can help review it.

alexeykudinkin commented on core/src/main/java/com/uber/m3/tally/sanitizers/StringSanitizer.java:
```diff
@@ -0,0 +1,35 @@
+ * StringSanitizer is to sanitize strings
+ * It has a Sanitize method which returns a sanitized version of the input string value.
+ */
+public interface StringSanitizer {
```
I don't think we need standalone interface for this, Function interface should be reflective

longquanzheng replied:
Yeah agreed. It's too heavy for a function to have interface. 
You meant functional interface/lambda, correct?

longquanzheng replied:
Changed to Function<String,String>

alexeykudinkin commented on core/src/main/java/com/uber/m3/tally/sanitizers/SanitizeRange.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * SanitizeRange is a range of characters (inclusive on both ends).
+ */
+public class SanitizeRange {
```
This is rather `CharRange`

alexeykudinkin commented on core/src/main/java/com/uber/m3/tally/sanitizers/SanitizeRange.java:
```diff
@@ -0,0 +1,47 @@
+
+    private SanitizeRange(char low, char high) {
+        this.low = low;
+        this.high = high;
```
Please add precondition asserting that `high >= low`

longquanzheng replied:
I don't see an example of "Precondition"(maybe I didn't find in a correct way) so I throw a Runtime exception here. LMK if you want to assert in a different way.

alexeykudinkin replied:
Yeah, precondition reference was mostly notional (please do not import Guava for that)

alexeykudinkin commented:
@longquanzheng we would also need to write a benchmark for this one.

longquanzheng commented:
@alexeykudinkin thanks so much for your reviews. I have addressed your comments except bench tests. Could you give me more about how to write/add it?

alexeykudinkin commented on core/src/main/java/com/uber/m3/tally/ScopeBuilder.java:
```diff
@@ -55,6 +57,7 @@
     protected String separator = DEFAULT_SEPARATOR;
     protected ImmutableMap<String, String> tags;
     protected Buckets defaultBuckets = DEFAULT_SCOPE_BUCKETS;
+    protected ScopeSanitizer sanitizer = new ScopeSanitizerBuilder().build();
```
Let's make this opt-in -- by default sanitizer should be no-op

longquanzheng replied:
Do you mean having it as `null` by default in the builder? 

This would be a little bit tricky. Then the ScopeImpl will have to do this null check everything, is that okay? (I would be easier if this is Kotlin and we can use `?` operator, sigh)

longquanzheng replied:
NVM, looks like I will apply another comment from you and Andrey in https://github.com/uber-java/tally/pull/97#discussion_r618596672

alexeykudinkin commented:
@longquanzheng LGTM, minor comment.

Regarding benchmarks, you can take a look at the benchmarks package for some examples;

SokolAndrey commented on core/src/main/java/com/uber/m3/tally/ScopeImpl.java:
```diff
@@ -136,12 +144,32 @@ public void close() {
+        }
+
+        ImmutableMap.Builder<String, String> builder = new ImmutableMap.Builder<>();
+        if (tags != null) {
```
nit: this check is redundant. If `tags` is null it will fail on `tags.entrySet()`

alexeykudinkin replied:
I'd rather re-phrase this: @longquanzheng please move this check at the beginning to make sure there's no NPE

longquanzheng replied:
yeah good point. Thanks

SokolAndrey commented on core/src/main/java/com/uber/m3/tally/sanitizers/CharRange.java:
```diff
@@ -0,0 +1,50 @@
+
+    private CharRange(char low, char high) {
+        if (low > high) {
+            throw new RuntimeException("invalid CharRange");
```
nit: s/RuntimException/IllegalArgumentException?

alexeykudinkin replied:
+1

longquanzheng replied:
Done. Thanks

SokolAndrey commented on core/src/main/java/com/uber/m3/tally/sanitizers/SanitizerImpl.java:
```diff
@@ -0,0 +1,69 @@
+    private final Function<String, String> keySanitizer;
+    private final Function<String, String> valueSanitizer;
+
+    SanitizerImpl(Function<String, String> nameSanitizer, Function<String, String> keySanitizer, Function<String, String> valueSanitizer) {
```
nit: I'd make arguments name more explicit: `tagKeySanitizer`, `tagValueSanitizer`

longquanzheng replied:
yeah good catch -- I should have done it, hehe

SokolAndrey commented on core/src/main/java/com/uber/m3/tally/sanitizers/ScopeSanitizerBuilder.java:
```diff
@@ -0,0 +1,89 @@
+ */
+public class ScopeSanitizerBuilder {
+
+    private Function<String, String> nameSanitizer = value -> value;
```
nit: consider `Function.identity()`

longquanzheng replied:
Thanks! TIL

SokolAndrey commented on core/src/main/java/com/uber/m3/tally/sanitizers/ScopeSanitizerBuilder.java:
```diff
@@ -0,0 +1,89 @@
+
+/**
+ * The SanitizerBuilder returns a Sanitizer for the name, key and value. By
+ * default, the name, key and value sanitize functions returns all the input
```
Maybe create a `NoopSanitizer implements ScopeSanitizer` and use it as a default one instead?

alexeykudinkin replied:
+1

longquanzheng replied:
Yeah I like this idea better. 
Otherwise having null in ScopeImpl will require lots of null checking in the code.

SokolAndrey commented on core/src/main/java/com/uber/m3/tally/sanitizers/ScopeSanitizerBuilder.java:
```diff
@@ -0,0 +1,89 @@
+
+    public ScopeSanitizer build() {
+        if (nameCharacters != null) {
+            nameSanitizer = nameCharacters.sanitizeStringFunc(repChar);
```
if both `nameCharcters` and `nameSanitizer` provided (same for tag's keys and values) this will override the sanitizer and use sanitize function from `ValidCharacters`. If this is intentional behavior I think it should be documented more explicit. Or maybe throw an exception if someone tries to provide a sanitizer function when characters are already set?

longquanzheng replied:
good catch! I like throwing exception. Less document is better.

longquanzheng replied:
also added new tests for them.

SokolAndrey commented on core/src/main/java/com/uber/m3/tally/sanitizers/CharRange.java:
```diff
@@ -0,0 +1,50 @@
+    char low() {
+        return low;
+    }
+
```
consider adding something like `public boolean isWithinRange(char ch){...}`

longquanzheng replied:
ok

SokolAndrey commented on core/src/main/java/com/uber/m3/tally/sanitizers/ValidCharacters.java:
```diff
@@ -0,0 +1,110 @@
+    public static final List<Character> UNDERSCORE_DASH_DOT_CHARACTERS = Arrays.asList('_', '-', '.');
+
+    private final List<CharRange> ranges;
+    private final List<Character> characters;
```
consider using `Set` instead of `List`

alexeykudinkin replied:
+1

SokolAndrey commented on core/src/main/java/com/uber/m3/tally/sanitizers/ValidCharacters.java:
```diff
@@ -0,0 +1,110 @@
+                // first check if the provided character is valid
+                boolean validCurr =
+                    ranges.stream().anyMatch(range -> ch >= range.low() && ch <= range.high())
+                        || characters.stream().anyMatch(character -> character.equals(ch));
```
`.contains` seems more readable to me and more efficient if you use `Set`

longquanzheng replied:
Good idea. will do it

SokolAndrey commented on core/src/main/java/com/uber/m3/tally/sanitizers/ScopeSanitizerBuilder.java:
```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
```
pls update copyright to 2021

longquanzheng replied:
done

SokolAndrey commented on core/src/test/java/com/uber/m3/tally/sanitizers/SanitizeRangeTest.java:
```diff
@@ -0,0 +1,40 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class SanitizeRangeTest {
```
nit: s/SanitizeRangeTest/CharRangeTest

SokolAndrey commented on core/src/test/java/com/uber/m3/tally/sanitizers/SanitizeRangeTest.java:
```diff
@@ -0,0 +1,40 @@
+    private static final char HIGH = 'z';
+
+    @Test
+    public void sanitizeRange() {
```
pls add a test when HIGH is smaller than LOW and maybe when HIGH == LOW

SokolAndrey commented on core/src/test/java/com/uber/m3/tally/sanitizers/SanitizerTest.java:
```diff
@@ -0,0 +1,108 @@
+
+import static org.junit.Assert.assertEquals;
+
+public class SanitizerTest {
```
nit: s/SanitizerTest/SanitizerImplTest or ScopeSanitizerTest

longquanzheng replied:
changed to ScopeSanitizerTest

SokolAndrey commented on m3/src/main/java/com/uber/m3/tally/m3/M3Sanitizer.java:
```diff
@@ -0,0 +1,50 @@
+            .withKeyCharacters(
+                ValidCharacters.of(
+                    ValidCharacters.ALPHANUMERIC_RANGE,
+                    ValidCharacters.UNDERSCORE_DASH_CHARACTERS))
```
iirc dot is also a valid character but not recommended

longquanzheng replied:
changed to UNDERSCORE_DASH_CHARACTERS

SokolAndrey reviewed (approved):
Since you're modifying `ScopeImpl` default behaviour (even though it's a noop by default) would be nice to see how does it affect the [benchmarks](https://github.com/uber-java/tally/blob/master/core/src/jmh/java/com/uber/m3/tally/ScopeImplBenchmark.java). 
You can run those benchmark tests using the following command:
```
./gradlew tally-core:runJmhTests
```
Could you run it without and with your changes and add both results to the PR?
It would be nice to have separate benchmarks for sanitizer implementation though, but I don't think  it has to be a part of this PR. @alexeykudinkin  wdyt?

you can find more info/examples on JMH benchmarks [here](https://github.com/openjdk/jmh/tree/master/jmh-samples/src/main/java/org/openjdk/jmh/samples)

longquanzheng commented:
Thanks both for the review. I will address the comments this week.

longquanzheng commented:
@alexeykudinkin @SokolAndrey thank you so much again for the reviewing. 

I have addressed all the comments. Below is the bench results on my laptop:

Before the PR(using master)
```
Benchmark                                    Mode  Cnt     Score    Error   Units
ScopeImplBenchmark.scopeReportingBenchmark  thrpt   10  1162.629 ± 67.777  ops/ms
```

After the PR:
```
Benchmark                                    Mode  Cnt     Score    Error   Units
ScopeImplBenchmark.scopeReportingBenchmark  thrpt   10  1146.372 ± 30.429  ops/ms

```

alexeykudinkin commented:
@longquanzheng i assume these runs are w/ no-op sanitizer? Can you also paste Benchmarks w/ the sanitizer actually used?

longquanzheng commented:
> @longquanzheng i assume these runs are w/ no-op sanitizer? Can you also paste Benchmarks w/ the sanitizer actually used?

Yes. Thanks for pointing it out. 
I am testing with sanitizer now. Is that the right way? https://github.com/longquanzheng/tally/commit/5f2993c5e575c00b384f6a17ddc5a58b98c6845d

longquanzheng commented:
> > @longquanzheng i assume these runs are w/ no-op sanitizer? Can you also paste Benchmarks w/ the sanitizer actually used?
> 
> Yes. Thanks for pointing it out.
> I am testing with sanitizer now. Is that the right way? [longquanzheng@5f2993c](https://github.com/longquanzheng/tally/commit/5f2993c5e575c00b384f6a17ddc5a58b98c6845d)

Just ran with the commit of sanitizer:
```
Benchmark                                    Mode  Cnt     Score    Error   Units
ScopeImplBenchmark.scopeReportingBenchmark  thrpt   10  1485.831 ± 40.634  ops/ms
```

alexeykudinkin commented:
@longquanzheng i don't think we have any Benchmarks before targeting specifically code paths you're changing. Let's make sure that Benchmarks are stressing code paths that are changing.

longquanzheng commented:
> @longquanzheng i don't think we have any Benchmarks before targeting specifically code paths you're changing. Let's make sure that Benchmarks are stressing code paths that are changing.

I got what you mean. Just added benchmark code. Can you take another look?

longquanzheng commented:
@alexeykudinkin here is the results:
```
Benchmark                                                 Mode  Cnt   Score   Error   Units
ScopeImplBenchmark.recordingBenchmark                    thrpt   10  75.436 ± 2.179  ops/ms
ScopeImplBenchmark.recordingWithSanitizingDashBenchmark  thrpt   10  53.200 ± 1.539  ops/ms

```

alexeykudinkin commented on core/benchmark-tests.txt:
```diff
@@ -0,0 +1,4 @@
+Benchmark                                                 Mode  Cnt      Score     Error   Units
+ScopeImplBenchmark.recordingBenchmark                    thrpt   10     74.819 ±   4.573  ops/ms
```
Please check how other benchmarks are reported (we need to capture all data-points offered by JMH)

longquanzheng replied:
Sorry, I don't really understand how exactly to do it. 

Would you mind taking it over? I am not convinced that this should be done by this PR, as the benchmark is already missing. The most problem is I don't have a clear picture of what you really want for benchmarking, and it should be much easier if your team just do it. 

LMK if you agree.

alexeykudinkin replied:
It should be done by this PR. This is a critical library laying in the hot-path of execution of many services, hence we need to maintain the focus on its performance. 

Over the last year a lot of efforts have been put in to optimize, streamline, and make this library robust to serve the needs it was originally built for. Therefore, as a new contribution guideline we require any non-trivial change to adhere to the same basic principles of assuring library's correctness and performance.

Totally appreciate the amount of incremental effort to require to adhere to this heightened standards of every contribution, but unfortunately there's no other way to guarantee high-level of reliability and performance of the open-source library otherwise.

longquanzheng replied:
> Please check how other benchmarks are reported

Can you give more details here? I did check but don't know how that works(though I see other reports have the details like gc time). 
For example another benchmark is just
`public class M3ReporterBenchmark extends AbstractReporterBenchmark`. So I don't know what am I missing here.

alexeykudinkin replied:
@SokolAndrey can you help in here?

alexeykudinkin commented on core/src/jmh/java/com/uber/m3/tally/ScopeImplBenchmark.java:
```diff
@@ -67,28 +62,56 @@
+                    .reportEvery(Duration.MAX_VALUE);
+        }
 
+        public void recordTestMetrics(final ScopeImpl scope) {
```
I don't think this is a good routine to check -- this was used as pre-init seq, and i would suggest to keep it as such.

Instead create separate benchmarks and routines to update counter/gauge/histogram separately.

longquanzheng replied:
Same as my pervious response. I appreciate your time reviewing and commenting it. 
But it would be nice if you can help the benchmarking.

alexeykudinkin replied:
Please check my comment above. We're more than happy to help w/ guidance, but benchmarking is now a standard requirement from non-trivial contributions.

longquanzheng replied:
Sure, will revert that one back. 

Can you give me details about what you want to benchmark? -- which method/function/routines?
